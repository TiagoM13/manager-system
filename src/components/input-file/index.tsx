import React from 'react';
import { Control, Controller, FieldError } from 'react-hook-form';

import { CircleNotch, UploadSimple } from '@phosphor-icons/react';

import avatarImageUrl from '@/assets/avatars/avatar-woman.jpeg';

import { Avatar } from '../avatar';

interface InputFileProps {
  placeholder?: string;
  loading?: boolean;
  hasPreview?: boolean;
  control: Control<any>;
  name: any;
  error?: FieldError | undefined;
}

export const InputFile: React.FC<InputFileProps> = ({
  placeholder = 'Escolher foto',
  loading = false,
  hasPreview = false,
  control,
  error,
  name,
}) => {
  const [imageUrl, setImageUrl] = React.useState<string | null>(null);

  const handleFileSelected = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { files } = event.target;

    if (!files) {
      return;
    }

    const previwURL = URL.createObjectURL(files[0]);

    setImageUrl(previwURL);
  };

  const defaultAvatar =
    'https://images.nightcafe.studio/jobs/0Kn1rxnTPMBeajQI9rVD/0Kn1rxnTPMBeajQI9rVD--2--bfk8i_2x.jpg?tr=w-1600,c-at_max';

  return (
    <div className="flex items-center gap-5 mt-4">
      {hasPreview && (
        <Avatar
          className="size-32"
          color="dark"
          imageUrl={imageUrl || defaultAvatar}
          name="avatar"
        />
      )}

      <label
        htmlFor="image_url"
        className={`bg-sky-600 flex items-center gap-3 py-2 px-4 rounded-md transition-all ease-in duration-500 focus-visible:bg-sky-600 text-sm font-medium text-white ${loading ? 'bg-sky-700' : 'hover:bg-sky-500 cursor-pointer'}`}
      >
        {loading ? (
          <CircleNotch
            weight="bold"
            color="white"
            className="size-5 animate-spin"
          />
        ) : (
          <UploadSimple className="size-5 text-white" weight="bold" />
        )}

        {placeholder}
      </label>

      <Controller
        name={name}
        control={control}
        render={({ field: { onChange } }) => (
          <input
            id="image_url"
            name={name}
            type="file"
            accept=".png,.jpg"
            className="hidden"
            disabled={loading}
            onChange={(e) => {
              onChange(e);
              handleFileSelected(e);
            }}
          />
        )}
      />

      {!!error && (
        <span className="block text-sm text-red-500 mt-1">{error.message}</span>
      )}
    </div>
  );
};
