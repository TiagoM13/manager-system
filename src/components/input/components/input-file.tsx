import React from 'react';
import { Control, Controller, FieldError } from 'react-hook-form';

import { CircleNotch, UploadSimple } from '@phosphor-icons/react';

import { ErrorMessage } from '@/components';

interface FileUploadButtonProps {
  placeholder?: string;
  defaultValue?: string;
  loading?: boolean;
  disabled?: boolean;
  control?: Control;
  error?: FieldError | undefined;
  name: string;
  onChangeFileSelected: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

export const FileUploadInput: React.FC<FileUploadButtonProps> = ({
  placeholder = 'Escolher foto',
  loading = false,
  disabled = false,
  defaultValue,
  onChangeFileSelected,
  control,
  error,
  name,
}) => {
  return (
    <label
      htmlFor={name}
      className={`bg-sky-600 flex items-center gap-3 py-2 px-4 rounded-md transition-all ease-in duration-500 focus-visible:bg-sky-600 text-sm font-medium text-white ${loading || disabled ? 'bg-sky-700' : 'hover:bg-sky-500 cursor-pointer'}`}
    >
      {loading ? (
        <CircleNotch
          data-testid="loading-icon"
          weight="bold"
          color="white"
          className="size-5 animate-spin"
        />
      ) : (
        <UploadSimple
          data-testid="upload-icon"
          className="size-5 text-white"
          weight="bold"
        />
      )}

      {placeholder}

      {control ? (
        <Controller
          name={name}
          control={control}
          defaultValue={defaultValue}
          render={({ field: { onChange } }) => (
            <input
              id={name}
              name={name}
              type="file"
              accept="image/*"
              className="hidden"
              disabled={loading || disabled}
              onChange={(e) => {
                onChange(e);
                onChangeFileSelected(e);
              }}
            />
          )}
        />
      ) : (
        <input
          id={name}
          name={name}
          type="file"
          accept="image/*"
          className="hidden"
          disabled={loading || disabled}
          onChange={onChangeFileSelected}
        />
      )}

      <ErrorMessage error={error?.message} />
    </label>
  );
};
