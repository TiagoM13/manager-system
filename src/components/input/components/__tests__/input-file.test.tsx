import React from 'react';
import '@testing-library/jest-dom';

import { fireEvent, render } from '@testing-library/react';

import { FileUploadInput } from '../input-file';

describe('<FileUploadInput/>', () => {
  test('should render currectly component', () => {
    const { getByText, getByTestId } = render(
      <FileUploadInput name="fileUpload" onChangeFileSelected={jest.fn()} />,
    );

    const icon = getByTestId('upload-icon');

    expect(getByText(/Escolher foto/i)).toBeInTheDocument();
    expect(icon).toBeInTheDocument();
  });

  test('should disable the buton when loading', () => {
    const { container } = render(
      <FileUploadInput
        name="fileUpload"
        onChangeFileSelected={jest.fn()}
        loading
      />,
    );

    const element = container.querySelector('.animate-spin');
    expect(element).toBeInTheDocument();
  });

  test('should call onChangeFileSelected when a file is selected', () => {
    const mockOnChangeFileSelected = jest.fn();

    const { asFragment, getByLabelText } = render(
      <FileUploadInput
        name="fileUpload"
        onChangeFileSelected={mockOnChangeFileSelected}
      />,
    );

    const inputElement = getByLabelText(/escolher foto/i);

    fireEvent.change(inputElement, {
      target: {
        file: [new File(['file'], 'test-file.jpg', { type: 'image/jpeg' })],
      },
    });

    expect(mockOnChangeFileSelected).toHaveBeenCalledTimes(1);
    expect(asFragment()).toMatchSnapshot();
  });
});
