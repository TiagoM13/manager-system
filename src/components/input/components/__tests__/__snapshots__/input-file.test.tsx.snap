// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileUploadInput/> should call onChangeFileSelected when a file is selected 1`] = `
<DocumentFragment>
  <label
    class="bg-sky-600 flex items-center gap-3 py-2 px-4 rounded-md transition-all ease-in duration-500 focus-visible:bg-sky-600 text-sm font-medium text-white hover:bg-sky-500 cursor-pointer"
    for="fileUpload"
  >
    <svg
      class="size-5 text-white"
      data-testid="upload-icon"
      fill="currentColor"
      height="1em"
      viewBox="0 0 256 256"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M228,144v64a12,12,0,0,1-12,12H40a12,12,0,0,1-12-12V144a12,12,0,0,1,24,0v52H204V144a12,12,0,0,1,24,0ZM96.49,80.49,116,61v83a12,12,0,0,0,24,0V61l19.51,19.52a12,12,0,1,0,17-17l-40-40a12,12,0,0,0-17,0l-40,40a12,12,0,1,0,17,17Z"
      />
    </svg>
    Escolher foto
    <input
      accept="image/*"
      class="hidden"
      id="fileUpload"
      name="fileUpload"
      type="file"
    />
  </label>
</DocumentFragment>
`;
